### PR TITLE
Use CTC loss as an auxiliary loss for transducer training.

### PR DESCRIPTION
**Pull request by csukuangfj**

### CODE REVIEW

File: egs/librispeech/ASR/pruned_transducer_stateless8/model.py [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+                dtype=torch.int64,
+            )
+
+            target_lengths = torch.tensor(
+                target_lengths,
+                device=x.device,
+                dtype=torch.int64,
+            )
+
+            ctc_loss = torch.nn.functional.ctc_loss(
```
Comment (csukuangfj) [inline egs/librispeech/ASR/pruned_transducer_stateless8/model.py:223]: Note that I am using torch CTC loss here.

I suggest using k2 CTC loss only when a word has multiple pronunciations.

File: egs/librispeech/ASR/pruned_transducer_stateless8/train.py [hunk 2]
```diff
@@ -0,0 +1,1152 @@
...
+  --world-size 4 \
+  --num-epochs 30 \
+  --start-epoch 1 \
+  --exp-dir pruned_transducer_stateless5/exp \
+  --full-libri 1 \
+  --max-duration 300
+
+# For mix precision training:
+
+./pruned_transducer_stateless5/train.py \
```
Comment (danpovey) [inline egs/librispeech/ASR/pruned_transducer_stateless8/train.py:35]: says 5

File: egs/librispeech/ASR/pruned_transducer_stateless8/model.py [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+                target_lengths.append(len(t))
+                targets.extend(t)
+
+            targets = torch.tensor(
+                targets,
+                device=x.device,
+                dtype=torch.int64,
+            )
+
+            target_lengths = torch.tensor(
```
Comment (kobenaxie) [inline egs/librispeech/ASR/pruned_transducer_stateless8/model.py:217]: Actually, `targets==y_padded` and `target_lengths==y_lens`.

Comment (csukuangfj) [reply egs/librispeech/ASR/pruned_transducer_stateless8/model.py:217]: `targets` is a 1-D tensor here while y_padded is a 2-D tensor.

Yes, you're right. `target_lengths == y_lens`.

**Comment by csukuangfj** [comment]
From https://lhotse.readthedocs.io/en/latest/kaldi.html
```python3
# Convert data/train to train_manifests/{recordings,supervisions}.json
lhotse kaldi import \
    data/train \
    16000 \
    train_manifests

# Convert train_manifests/{recordings,supervisions}.json to data/train
lhotse kaldi export \
    train_manifests/recordings.json \
    train_manifests/supervisions.json \
    data/train
```

File: egs/librispeech/ASR/pruned_transducer_stateless8/model.py (hunk 3 above)
Comment (kobenaxie) [reply egs/librispeech/ASR/pruned_transducer_stateless8/model.py:217]: oh, get it, The 1-D tensor `targets` will save some memory  than 2-D tensor `y_padded`.

**Comment by csukuangfj** [comment]
Here are the results of this PR.

I tried two models:
- **exp-L**: It is from https://github.com/k2-fsa/icefall/blob/master/egs/librispeech/ASR/RESULTS.md#large, which has about 118 M parameters
- **exp**: It is from https://github.com/k2-fsa/icefall/blob/master/egs/librispeech/ASR/RESULTS.md#baseline-2, which has about 88 M parameters

It shows:
- Add CTC loss as an auxiliary loss in RNN-T training is helpful
- ctc loss scale 0.20 is a good value for both settings


---

## Results for exp-L
The WERs in the cell are for `test_clean/test_other`
| ctc_loss_scale | greedy_search | modified_beam_search | fast_beam_search | comment         |
|----------------|---------------|----------------------|------------------|-----------------|
| 0.0 (baseline) | 2.39/5.57     | 2.35/5.50            | 2.38/5.50        | epoch-39-avg-7  |
| 0.10           | 2.40/5.29     | 2.39/**5.26**            | 2.39/5.27        | epoch-30-avg-14 |
| 0.20           | 2.29/5.36     | **2.27**/5.35            | 2.28/5.31        | epoch-38-avg-23 |
| 0.25           | 2.31/5.52     | 2.28/5.52            | 2.30/5.53        | epoch-30-avg-15 |
| 0.30           | 2.33/5.40     | 2.30/5.35            | 2.32/5.34        | epoch-37-avg-21 |
| 0.40           | 2.32/5.38     | 2.28/5.37            | 2.33/5.37        | epoch-40-avg-28 |
| 0.50           | 2.33/5.55     | 2.32/5.55            | 2.35/5.55        | epoch-29-avg-14 |

## Results for exp
| ctc_loss_scale | greedy_search | modified_beam_search | fast_beam_search | comment         |
|----------------|---------------|----------------------|------------------|-----------------|
| 0.0 (baseline) | 2.41/5.70     | 2.41/5.69            | 2.41/5.69        | epoch-31-avg-17 |
| 0.10           | 2.41/5.59     | 2.42/**5.55**            | 2.42/5.57        | epoch-39-avg-25 |
| 0.20           | 2.34/5.60     | **2.33**/5.59            | 2.34/5.56        | epoch-38-avg-16 |
| 0.25           | 2.43/5.72     | 2.39/5.67            | 2.39/5.66        | epoch-37-avg-20 |
| 0.30           | 2.42/5.63     | 2.40/5.57            | 2.40/5.57        | epoch-31-avg-13 |
| 0.40           | 2.37/5.67     | 2.33/5.66            | 2.34/5.66        | epoch-39-avg-23 |
| 0.50           | 2.39/5.85     | 2.40/5.82            | 2.39/5.82        | epoch-29-avg-13       |

**Comment by yaozengwei** [comment]
I use the above trained models for **ctc decoding** and also get promising results. 

<div class="okr-block-clipboard" data-okr="%7B%22okrDelta%22%3A%5B%7B%22lineType%22%3A%22unsupport%22%2C%22lineOptions%22%3A%7B%7D%2C%22lineContent%22%3A%5B%5D%7D%5D%2C%22businessKey%22%3A%22lark-doc%22%7D"></div><div data-zone-id="0" data-line-index="0" style="white-space: pre;">

ctc-loss-scale | ctc-decoding | 1best | Nbest | Nbest-rescoring | whole-lattice-rescoring | comment
-- | -- | -- | -- | -- | -- | --
0.10 | 2.66 & 5.92 | 2.71 & 5.4 | 2.71 & 5.39 | 2.69 & 5.3 | 2.68 & **5.29** | epoch-40-avg-14
0.20 | 2.54 & 5.87 | 2.67 & 5.6 | 2.67 & 5.6 | 2.61 & 5.45 | 2.6 & 5.45 | epoch-40-avg-20
0.25 | 2.54 & 5.78 | 2.72 & 5.45 | 2.72 & 5.45 | 2.66 & **5.29** | 2.65 & **5.29** | epoch-40-avg-18
0.30 | 2.52 & 5.79 | 2.66 & 5.47 | 2.66 & 5.47 | 2.59 & 5.33 | 2.58 & 5.34 | epoch-40-avg-14
0.40 | **2.5** & 5.89 | 2.69 & 5.57 | 2.69 & 5.57 | 2.63 & 5.46 | 2.63 & 5.52 | epoch-40-avg-10

</div>
